File: Heuristic Search (sgware)/src/edu/uno/ai/planning/h/HeuristicQueue.java
package edu.uno.ai.planning.h;

import java.util.Comparator;
import java.util.PriorityQueue;

import edu.uno.ai.planning.ss.StateSpaceNode;

public class HeuristicQueue {
	
	public final StateHeuristic heuristic;
	private final PriorityQueue<HeuristicNode> queue;
	
	public HeuristicQueue(StateHeuristic heuristic, HeuristicComparator comparator) {
		this.heuristic = heuristic;
		this.queue = new PriorityQueue<>(new Comparator<HeuristicNode>(){
			@Override
			public int compare(HeuristicNode n1, HeuristicNode n2) {
				double comparison = comparator.compare(n1, n2);
				if(comparison == 0)
					return n1.id - n2.id;
				else if(comparison < 0)
					return -1;
				else
					return 1;
			}
		});
	}
	
	public void clear() {
		queue.clear();
	}
	
	public int size() {
		return queue.size();
	}
	
	public HeuristicNode push(StateSpaceNode node) {
		HeuristicNode n = new HeuristicNode(node, heuristic.evaluate(node.state));
		queue.add(n);
		return n;
	}
	
	public HeuristicNode peek() {
		return queue.peek();
	}
	
	public double hPeek() {
		return queue.peek().heuristic;
	}
	
	public StateSpaceNode pop() {
		return queue.poll().state;
	}
}

File: UnitySokoban/Assets/Scripts/Planning/HeuristicSearchPlannerSGW/HeuristicQueue.cs
using Planning.Util;
using StateSpaceSearchProject;
using System;
using System.Collections.Generic;

namespace HeuristicSearchPlannerSGW
{
    public class HeuristicQueue
    {
        public readonly StateHeuristic heuristic;
        private readonly PriorityQueue<HeuristicNode> queue;

        public HeuristicQueue(StateHeuristic heuristic, HeuristicComparator comparator)
        {
            this.heuristic = heuristic;
            this.queue = new PriorityQueue<HeuristicNode>();
        }

        public void clear()
        {
            queue.Clear();
        }

        public int size()
        {
            return queue.Size();
        }

        public HeuristicNode push(StateSpaceNode node)
        {
            int heuristicValue = heuristic.evaluate(node.state);
            if (heuristicValue != HSPHeuristic.INFINITY)
                heuristicValue += node.plan.Size();
            HeuristicNode n = new HeuristicNode(node, heuristicValue);
            queue.Enqueue(n, heuristicValue);
            return n;
        }

        public HeuristicNode peek()
        {
            return queue.Peek();
        }

        public double hPeek()
        {
            return queue.Peek().heuristic;
        }

        public StateSpaceNode pop()
        {
            return queue.Dequeue().state;
        }
    }
}